### PR TITLE
feat(KaryaKarsa): add presence

### PR DESCRIPTION
**Pull request by kyuiki**
This is how the presence work.

- At Feeds / HomePage
DiscordPTB_hcWdCYE1RF

- At Discover
DiscordPTB_rX0ABSuMP9

- At Discover > Search
DiscordPTB_k6BfSsqXhH

- At Dashboard (The state will change depend on subpages)
DiscordPTB_YNh2UA6jI4

- At User Profile
0WimZ4CJBt

- At Someone Page
DiscordPTB_4cfQCGZzVh

- At Someone Post
DiscordPTB_gDA4nwShFF

- Default
DiscordPTB_SEWGy47ApZ

### CODE REVIEW

**Review by Slowlife01 — CHANGES_REQUESTED** [review]
Revert your change to yarn.lock and format your code with prettier

**Comment by QkieMauln** [comment]
> Revert your change to yarn.lock and format your code with prettier

hmm ok, i just hard reset-ed the stuff

File: websites/K/KaryaKarsa/presence.ts [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    if (!b && !!paths[2]) return p.name === "*";
+    return b;
+  }).state;
+};
+
+presence.on("UpdateData", () => {
+const presenceData: PresenceData = {
+  largeImageKey: "main",
+};
+presenceData.startTimestamp = timestamp;
```
Comment (Slowlife01) [inline websites/K/KaryaKarsa/presence.ts:134]:
```suggestion
const presenceData: PresenceData = {
  largeImageKey: "main",
  startTimestamp: timestamp
};
```

And merge the consts

File: websites/K/KaryaKarsa/presence.ts [hunk 2]
```diff
@@ -0,0 +1,231 @@
+const presence = new Presence({
+  clientId: "920135248006754394",
```
Comment (Slowlife01) [inline websites/K/KaryaKarsa/presence.ts:2]: No trailing comma

**Review by Slowlife01 — DISMISSED** [review]

File: websites/K/KaryaKarsa/presence.ts (hunk 1 above)
Comment (QkieMauln) [reply websites/K/KaryaKarsa/presence.ts]: alright

File: websites/K/KaryaKarsa/presence.ts (hunk 2 above)
Comment (QkieMauln) [reply websites/K/KaryaKarsa/presence.ts]: got it

File: websites/K/KaryaKarsa/presence.ts [hunk 3]
```diff
@@ -0,0 +1,231 @@
+const presence = new Presence({
+    clientId: "920135248006754394"
+  }),
+  timestamp = Math.floor(Date.now() / 1000),
```
Comment (EncryptedDev) [inline websites/K/KaryaKarsa/presence.ts:4]:
```suggestion
  browsingTimestamp = Math.floor(Date.now() / 1000),
```

**Review by EncryptedDev — CHANGES_REQUESTED** [review]

File: websites/K/KaryaKarsa/presence.ts (hunk 3 above)
Comment (QkieMauln) [reply websites/K/KaryaKarsa/presence.ts]: alrighty

File: websites/K/KaryaKarsa/presence.ts [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+      break;
+    case "feed":
+      presenceData.details = "Looking at Feeds";
+      break;
+    case "profile":
+      presenceData.details = "Looking at Profile";
+      break;
+    case "about" || "terms":
+      presenceData.details = "Looking at Terms of Service";
+      break;
```
Comment (Slowlife01) [inline websites/K/KaryaKarsa/presence.ts:150]:
```suggestion
    case "about":
    case "terms":
      presenceData.details = "Looking at Terms of Service";
      break;
```

Comment (QkieMauln) [reply websites/K/KaryaKarsa/presence.ts]: i see, alright

File: websites/K/KaryaKarsa/presence.ts [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    }).state;
+  };
+
+presence.on("UpdateData", () => {
+  const presenceData: PresenceData = {
+    largeImageKey: "main",
+    startTimestamp: browsingTimestamp
+  };
+
+  const paths: string[] = window.location.pathname.split("/").splice(1);
```
Comment (EncryptedDev) [inline websites/K/KaryaKarsa/presence.ts:136]: Resolve Lint annotation here

**Review by EncryptedDev — CHANGES_REQUESTED** [review]

File: websites/K/KaryaKarsa/presence.ts (hunk 5 above)
Comment (QkieMauln) [reply websites/K/KaryaKarsa/presence.ts:136]: uhh, you mean combine the const?

File: websites/K/KaryaKarsa/presence.ts [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+  // This function act like switch function but for objects Array (yea its a lil bit complicated)
+  stateSetter = (
+    paths: string[],
+    arrayPath: Array<{
+      name: string;
+      state?: string;
+      paths?: Array<{
+        name?: string;
+        state?: string;
+      }>;
```
Comment (ImRodry) [inline websites/K/KaryaKarsa/presence.ts:118]:
```suggestion
    arrayPath: {
      name: string;
      state?: string;
      paths?: Array<{
        name?: string;
        state?: string;
      }[];
```

File: websites/K/KaryaKarsa/presence.ts [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+      },
+      {
+        name: "goal",
+        state: "My Goal"
+      },
+      {
+        name: "tiers",
+        paths: [
+          {
+            name: "",
```
Comment (ImRodry) [inline websites/K/KaryaKarsa/presence.ts:33]: Can't this be removed if it's only an empty string?

File: websites/K/KaryaKarsa/presence.ts [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+      }>;
+    }>
+  ): string => {
+    const result = arrayPath.find((p) => p.name === (paths[1] || ""));
+    if (result.state) return result.state;
+    return result.paths.find((p) => {
+      const b: boolean = p.name === (paths[2] || "");
+      if (!b && !!paths[2]) return p.name === "*";
+      return b;
+    }).state;
```
Comment (ImRodry) [inline websites/K/KaryaKarsa/presence.ts:127]: Use nullish coalescing and avoid unnecessary type assertions
```suggestion
    const result = arrayPath.find((p) => p.name === (paths[1] ?? ""));
    return result.state ?? result.paths.find(p => {
      const b = p.name === (paths[2] ?? "");
      if (!b && paths[2]) return p.name === "*";
      return b;
    }).state;
```

File: websites/K/KaryaKarsa/presence.ts [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+presence.on("UpdateData", () => {
+  const presenceData: PresenceData = {
+    largeImageKey: "main",
+    startTimestamp: browsingTimestamp
+  },
+  paths: string[] = window.location.pathname.split("/").splice(1);
+
+  switch (paths[0].toLowerCase()) {
+    case "":
+      presenceData.details = "Looking at HomePage";
```
Comment (ImRodry) [inline websites/K/KaryaKarsa/presence.ts:139]: Won't suggest on all lines but please avoid unnecessary capitalization in the middle of the sentence
```suggestion
      presenceData.details = "Looking at the home page";
```

File: websites/K/KaryaKarsa/presence.ts [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+      return b;
+    }).state;
+  };
+
+presence.on("UpdateData", () => {
+  const presenceData: PresenceData = {
+    largeImageKey: "main",
+    startTimestamp: browsingTimestamp
+  },
+  paths: string[] = window.location.pathname.split("/").splice(1);
```
Comment (ImRodry) [inline websites/K/KaryaKarsa/presence.ts:135]:
```suggestion
  paths = window.location.pathname.split("/").splice(1);
```

File: websites/K/KaryaKarsa/presence.ts [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+      presenceData.details = "Looking at Dashboard";
+      presenceData.state = stateSetter(paths, pathData.dashboard);
+      break;
+    case "discover": {
+      presenceData.details = "Looking at Discover / Search";
+      const searchBar = <HTMLInputElement>(
+        document.querySelector(
+          ".q-field.q-input.q-field--filled.q-field--square.q-field--float.q-validation-component input.q-field__native.q-placeholder"
+        )
+      );
```
Comment (ImRodry) [inline websites/K/KaryaKarsa/presence.ts:161]:
```suggestion
      const searchBar = document.querySelector<HTMLInputElement>(
          ".q-field.q-input.q-field--filled.q-field--square.q-field--float.q-validation-component input.q-field__native.q-placeholder"
        );
```

File: websites/K/KaryaKarsa/presence.ts [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+        presenceData.state = `Searching for "${searchBar.value}"`;
+      break;
+    }
+    case "user":
+      presenceData.details = "Looking at My Detail";
+      presenceData.state = stateSetter(paths, pathData.user);
+      break;
+    default: {
+      // Default (If the path isnt defined here. Will be replaced with this!)
+      presenceData.details = "Browsing Pages";
```
Comment (ImRodry) [inline websites/K/KaryaKarsa/presence.ts:172]: If you wanna do what the comment above says (replace if not set), you should be doing this
```suggestion
      presenceData.details ??= "Browsing Pages";
```

File: websites/K/KaryaKarsa/presence.ts [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+
+      // if no path detected, mean it on homepage
```
Comment (ImRodry) [inline websites/K/KaryaKarsa/presence.ts:174]: Please either improve the phrasing on these comments or remove them entirely

File: websites/K/KaryaKarsa/presence.ts [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+      // if user Looking at user page
+      {
+        const titleUser = <HTMLElement>(
+            document.querySelector(`${getUserHeader} .q-px-md.q-py-md div`)
+          ),
+          roleUser = <HTMLElement>(
+            document.querySelector(
+              `${getUserHeader} .q-px-md.q-py-md .text-caption.text-grey-7`
+            )
+          );
```
Comment (ImRodry) [inline websites/K/KaryaKarsa/presence.ts:186]:
```suggestion
        const titleUser = document.querySelector<HTMLElement>(
        `${getUserHeader} .q-px-md.q-py-md div`),
          roleUser = document.querySelector<HTMLElement>(
              `${getUserHeader} .q-px-md.q-py-md .text-caption.text-grey-7`);
```

File: websites/K/KaryaKarsa/presence.ts [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+          tagUser = <HTMLElement>(
+            document.querySelector(
+              "#q-app > div > div > div.q-page-container.main.q-pb-xl.bordered-desktop.post-page > div.post-date.text-body2.q-px-md.q-pt-md.q-pb-sm.post-content > a"
+            )
+          ),
+          postTitle = <HTMLElement>(
+            document.querySelector(
+              "#q-app > div > div > div.q-page-container.main.q-pb-xl.bordered-desktop.post-page > h1"
+            )
+          );
```
Comment (ImRodry) [inline websites/K/KaryaKarsa/presence.ts:214]:
```suggestion
        const titleUser = document.querySelector<HTMLElement>(getUserHeaderHome),
          tagUser = document.querySelector<HTMLElement>(
              "#q-app > div > div > div.q-page-container.main.q-pb-xl.bordered-desktop.post-page > div.post-date.text-body2.q-px-md.q-pt-md.q-pb-sm.post-content > a"
          ),
          postTitle = document.querySelector<HTMLElement>(
              "#q-app > div > div > div.q-page-container.main.q-pb-xl.bordered-desktop.post-page > h1"
          );
```

**Review by ImRodry — CHANGES_REQUESTED** [review]

File: websites/K/KaryaKarsa/presence.ts (hunk 15 above)
Comment (QkieMauln) [reply websites/K/KaryaKarsa/presence.ts]: Alright, got it

**Comment by ImRodry** [comment]
Please mark conversations as resolved if you have addressed them

File: websites/K/KaryaKarsa/presence.ts [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+        const titleUser = document.querySelector<HTMLElement>(
+            `${getUserHeader} .q-px-md.q-py-md div`
+          ),
+          roleUser = document.querySelector<HTMLElement>(
+            `${getUserHeader} .q-px-md.q-py-md .text-caption.text-grey-7`
+          );
+
+        if (titleUser && roleUser) {
+          presenceData.details = titleUser.innerText;
+          presenceData.state = roleUser.innerText;
```
Comment (EncryptedDev) [inline websites/K/KaryaKarsa/presence.ts:176]:
```suggestion
          presenceData.details = titleUser.textContent;
          presenceData.state = roleUser.textContent;
```

File: websites/K/KaryaKarsa/presence.ts [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+          tagUser = document.querySelector<HTMLElement>(
+            "#q-app > div > div > div.q-page-container.main.q-pb-xl.bordered-desktop.post-page > div.post-date.text-body2.q-px-md.q-pt-md.q-pb-sm.post-content > a"
+          ),
+          postTitle = document.querySelector<HTMLElement>(
+            "#q-app > div > div > div.q-page-container.main.q-pb-xl.bordered-desktop.post-page > h1"
+          );
+
+        if (titleUser && tagUser && postTitle) {
+          presenceData.details = `${titleUser.innerText} (${tagUser.innerText})`;
+          presenceData.state = postTitle.innerText;
```
Comment (EncryptedDev) [inline websites/K/KaryaKarsa/presence.ts:199]:
```suggestion
          presenceData.details = `${titleUser.textContent} (${tagUser.textContent})`;
          presenceData.state = postTitle.textContent;
```

**Review by EncryptedDev — CHANGES_REQUESTED** [review]

File: websites/K/KaryaKarsa/presence.ts (hunk 17 above)
Comment (QkieMauln) [reply websites/K/KaryaKarsa/presence.ts]: ahh ok

**Review by EncryptedDev — DISMISSED** [review]

**Review by Bas950 — APPROVED** [review]

**Review by EncryptedDev — APPROVED** [review]

**Pull request merged**